### DISCUSSION TITLE
Useless without Prisma?

### DISCUSSION THREAD

**Discussion by psnehanshu** [trpc/trpc#4232] [General]
A huge part of why tRPC is able to provide such high degree of type safety is because Prisma returns typed results. I like how Prisma will change the return type based on what data we request, e.g. If we query a model called `User`, and we include another associated model called `Profile`, Prisma will return a intersection type like this

```typescript
User & { Profile: Profile }
```

Due to this when a procedure returns this object, in the frontend we access the property with full type safety like this `user.profile.email`. For me that's a huge benefit of using tRPC which just goes away when not using Prisma.

My point is tRPC docs should explicitly mention this, and state that the full power of tRPC won't be realized without Prisma.

**Comment by mmkal** [comment]
I disagree. Anything with decent types will work well with trpc. My team use trpc in production for virtually everything we do, and don't use Prisma. We use `@slonik/typegen` which adds types to Postgres queries directly, without an ORM. It's similarly useful with any backend APIs or libraries that  provide types.

**Reply by psnehanshu** [reply]
I just skimmed through the Readme of that package. Looks like you will have to first define the types manually and then use that while calling `sql`, which is not much different than just typing using `as`. If I understood how this package works correctly, then that's too much work for it to be useful.

**Reply by sachinraja** [reply]
We are not debating which solution is the best. We shouldn't impose our personal opinions on all tRPC users. I'd rather convey that tRPC works well with all these solutions.

**Reply by mmkal** [reply]
> I just skimmed through the Readme of that package. Looks like you will have to first define the types manually and then use that while calling `sql`, which is not much different than just typing using `as`. If I understood how this package works correctly, then that's too much work for it to be useful.

No, the types are generated automatically!

Anyway I agree with @sachinraja - the point is, it's not at all useless without Prisma.

**Comment by Nick-Lucas** [comment]
You can write types for anything, it's just TypeScript.

If you're consuming external REST APIs, then you can use OpenAPI Codegen to create a typesafe client.

If you're using a DB there are many solutions, everything from parsing results with Zod, to safeql, to another ORM than Prisma, or a query builder like kysely.

If your endpoints are returning `any` then that's entirely on you as a developer.

**Reply by psnehanshu** [reply]
Zod is good, safeql forces you to define types yourself, kysely also forces you to define the type manually. OpenAPI Codegen is useful for calling external APIs (provided those APIs are OpenAPI complaint).

So it's just more work without Prisma.

**Reply by Nick-Lucas** [reply]
Prisma is a cool tool. ORMs at large exist because they solve these problems. 

But your question seems to hinge on all type systems being irrelevant because IO is difficult and without an in-language tool controlling migration of the IO shape "the language is useless". Just because some tools require a little more up front work to define (and validate) types doesn't invalidate the tool itself. Without tRPC the frontend wouldn't know the types implicitly, so it saves a lot of work in creating the middle layer.

**Reply by igalklebanov** [reply]
@pSnehanshu 

>  kysely also forces you to define the type manually.

Nobody's forcing you to do anything. 🤷 

You can use prisma as dev dependency for migrations/schemas and generate kysely types from your prisma schemas using https://github.com/valtyr/prisma-kysely.

You can also use https://github.com/RobinBlomberg/kysely-codegen.

**Comment by sachinraja** [comment]
We won't do this. Prisma is not the only solution for fully typed database queries.

**Comment by aprilmintacpineda** [comment]
You're assuming that TRPC was built for Prisma? LOL.

**Comment by Nick-Lucas** [comment]
I'm going to close this as any discussion has run its course and the premise has been thoroughly and constructively responded to
